lru.go

```diff
@@ -4,116 +4,205 @@
-	}
-	return c, nil
+// New creates an LRU cache. It accepts a maximum size limit
+// and a maximum byte limit. If either one of these are set 
+// to zero, then that limit will be considered unlimitted but 
+// if both are set to zero then an error will be returned.
+// Exceeding either of these limits while calling Add() will
+// result in the least recently used item being dropped from
+// the cache and the onRemove() function being called for
+// that item.
+func New(items int, bytes int64) (*Cache, error) {
+    if items <= 0 && bytes <= int64(0) {
+        return nil, errors.New("Must provide a positive bound on either item count or byte count.")
+    }
+    onRemove := func(val interface{}) {
+        return
+    }
+    c := &Cache{
+        maxBytes:   bytes,
+        bytes:      int64(0),
+        maxItems:   items,
+        evictList:  list.New(),
+        items:      make(map[interface{}]*list.Element),
+        onRemove:   onRemove,
+    }
+    return c, nil
 }
 
-// Purge is used to completely clear the cache
+// Purge will completely clear the cache of all key, value pairs.
 func (c *Cache) Purge() {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.evictList = list.New()
-	c.items = make(map[interface{}]*list.Element, c.size)
+    c.lock.Lock()
+    defer c.lock.Unlock()            
+    for c.Len() > 0 {
+        c.removeOldest()
+    }
+    c.items = nil
+    c.items = make(map[interface{}]*list.Element)
 }
 
-// Add adds a value to the cache.
-func (c *Cache) Add(key, value interface{}) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
+// Schedule routine purges. May be used to free up memory or 
+// ensure that the OnRemove function is called within a certain
+// amount of time.
+func (cache *Cache) ScheduleClear(d time.Duration) {
+    t := time.Tick(d)
+    for {
+        select {
+        case <-t:
+            cache.Purge()
+        }
+    }
+}
 
-	// Check for existing item
-	if ent, ok := c.items[key]; ok {
-		c.evictList.MoveToFront(ent)
-		ent.Value.(*entry).value = value
-		return
-	}
+func getByteCount(key interface{}) int64 {
+    var buf bytes.Buffer
+    enc := gob.NewEncoder(&buf)
+    err := enc.Encode(key)
+    if err != nil {
+    	log.Println(err)
+        return int64(0)
+    }
+    return int64(len(buf.Bytes()))
+}
+
+// Add adds a value to the cache. If the new value causes the
+// set maxItems or maxBytes values to be exceeded, then the
+// least recently used item will be popped from the cache and 
+// the onRemove() function will be called for that value.
+func (c *Cache) Add(key, value interface{}) {
+    c.lock.Lock()
+    defer c.lock.Unlock()
+
+    // Check for existing item
+    if ent, ok := c.items[key]; ok {
+        c.evictList.MoveToFront(ent)
+        ent.Value.(*entry).Value = value
+        return
+    }
+
+    // Add new item
+    ent := &entry{key, value}
+    entry := c.evictList.PushFront(ent)
+    c.items[key] = entry
+    c.bytes += getByteCount(value)
```

Comment:
@bmconklin  Awesome! Thanks for making these changes. One more suggestion: `unsafe.Sizeof` (https://golang.org/pkg/unsafe/#Sizeof) will be faster and probably more accurate. Problem though: it'll only return the size of the interface, not the underlying concrete type. One possibility is to define a Sizer interface:

```go
type Sizer interface {
    Size() uintptr
}
```

A type assertion will tell whether the type implements this interface. Alternatively, a function could be created (like `AddSizer` that takes a `Sizer` instead of `interface{}`) so you could skip the type assertion and avoid the error case.

One final comment: with a size restriction, Add can fail if the item by itself is too big for the cache (an edge case, I know). Maybe Add should return an error.

FYI I'm quite interested in where this is going. One thing I hope this package will address in the future (if not sooner) is lock contention. But that's another issue entirely.